[getsquared] iP

getsquared opened the pull request:
# Toothless
```
   /\_/\
  ( o.o )
   > ^ <
```
> Made of **100% void** 🐈‍⬛ 
## About
Toothless is a catbot that remembers your tasks with its singular braincell. It's
* text-based
* ~~easy~~ SUPER EASY to learn and use

## Installation
1. Download it from [here](https://github.com/getsquared/ip/releases)
2. Open your command prompt
3. Run it using `java -jar toothless.jar`

## Features
- [x] Add, mark and delete tasks, including setting date and time.
- [x] View all tasks and save locally to a txt file.
- [x] Search for tasks with a keyword.
- [ ] _GUI coming soon_!

---
If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public static void main(String[] args) {
    Toothless toothless = new Toothless("./data/tasklist.txt");
    toothless.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

getsquared committed:
Add Level 0. Rename, Greet, Exit

getsquared committed:
Add Level 1. Echo

getsquared committed:
Add Level 2. Add, List

getsquared committed:
Add Task class

getsquared committed:
Level 3. Mark as Done

getsquared committed:
Add Task Subclasses

getsquared committed:
Add Level 4. ToDo, Event, Deadline

getsquared committed:
Add Text UI Testing

getsquared committed:
Add DukeException Class

getsquared committed:
Add Level 5. Handle Errors

getsquared committed:
Level 6. Delete

getsquared committed:
Abstract TaskList from main for OOP

getsquared committed:
Add Level 7. Save and Load from hard disk

getsquared committed:
Add Level 8. Dates and Times

getsquared committed:
Fix OOP between Storage and TaskList

getsquared committed:
Abstract Parser and Ui from Duke

getsquared committed:
Merge remote-tracking branch 'origin/add-gradle-support'

getsquared committed:
Add a few JUnit tests for TaskList and Parser

getsquared committed:
Merge branch 'branch-A-JUnit'

getsquared committed:
Rename Duke to Toothless and create .jar

getsquared committed:
Add JavaDoc for all classes

getsquared committed:
Improve coding standard

getsquared committed:
Add Level 9. Find

getsquared committed:
Merge branch 'branch-A-JavaDoc'

getsquared committed:
Resolve merge conflicts

getsquared committed:
Resolve merge conflicts Level 9

Ella-e commented on src/main/java/toothless/parser/Parser.java:
```diff
@@ -0,0 +1,204 @@
+     */
+    public int validateListInput(String listInput, String command, int taskListSize) throws ToothlessException {
+        // split string by spaces
+        String[] markInputSplit = listInput.strip().split("\\s+");
```
Would it be better to not start the variable name with a verb as it is just the place for storage, not for performing the actual action?

Ella-e commented on src/main/java/toothless/parser/Parser.java:
```diff
@@ -0,0 +1,204 @@
+     * @return String array containing task description and /by field.
+     * @throws ToothlessException if user input is invalid or in the wrong format.
+     */
+    public String[] validateDeadlineInput(String deadlineInput) throws ToothlessException {
```
For the naming of the function, maybe the verb 'validate' is not so accurate in describing the task it does. 'Validate' might make people think this function is just doing the validation and would return some boolean value or void, but this function actually breaks the deadlineInput down and put the result into an array and returns this array.

Ella-e commented on src/main/java/toothless/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+    private String filePath;
+    private File file;
+
+    private static final DateTimeFormatter DATETIME_PARSE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Good to see the constant name is meaning and in the right format.

Ella-e commented on src/main/java/toothless/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     */
+    public void saveToStorage(ArrayList<Task> tasks) throws ToothlessException {
+        try {
+            FileWriter fw = new FileWriter(this.filePath, false);
```
Would it be better to consider using more meaningful name like 'fileWriter' instead of 'fw'?

Ella-e reviewed (commented):
Good to see the code in compliance with the coding standard and naming standard. Just some cosmetic advices to double check. Good Job!

getsquared committed:
Configure checkstyle and fix violations

getsquared committed:
Merge branch 'branch-A-CheckStyle'

getsquared committed:
Add GUI and link commands

getsquared committed:
Merge branch 'branch-Level-10'

CaptainKeqing commented on src/main/java/toothless/parser/Parser.java:
```diff
@@ -0,0 +1,183 @@
+public class Parser {
+    /**
+     * The main method that parses user input.
+     * @param userInput User input string to be parsed.
```
Should there be a newline separating the description of the javadoc with the parameter section? This would better adhere to the coding standards [here](https://se-education.org/guides/conventions/java/intermediate.html#comments)

I see this problem in your other javadocs as well

CaptainKeqing commented on src/main/java/toothless/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * Returns a String representing the task formatted for printing.
+     * @return String representing the task formatted for printing.
+     */
```
Just wanted to say good job on your javadoc coverage ! I see you put alot of effort into writing documentation that makes your code easy to understand :smile:

CaptainKeqing commented on src/main/java/toothless/parser/Parser.java:
```diff
@@ -0,0 +1,183 @@
+        String response = "";
+        if (userInput.equals("bye")) {
+            response = "Bye. Purr-lease chat again soon!";
+        } else if (userInput.equals("list")) {
+            response = ui.getListMessage(tasks.getTasks(),
+                    "Oops! Looks like you haven't added any tasks yet!",
+                    "Here are the tasks in your list:\n");
+        } else if (userInput.startsWith("mark ") || userInput.equals("mark")) {
+            int listIndex = validateListInput(userInput, "mark", tasks.size());
+            Task markedTask = tasks.markTask(listIndex);
+            response = ui.getMarkedTaskMessage(markedTask);
+        } else if (userInput.startsWith("unmark ") || userInput.equals("unmark")) {
+            int listIndex = validateListInput(userInput, "unmark", tasks.size());
+            Task unmarkedTask = tasks.unmarkTask(listIndex);
+            response = ui.getUnmarkedTaskMessage(unmarkedTask);
+        } else if (userInput.startsWith("todo ") || userInput.equals("todo")) {
+            String taskDescription = validateToDoInput(userInput);
+            Task newTask = tasks.addToDoToList(taskDescription);
+            response = ui.getNewTaskMessage(newTask, tasks.size());
+        } else if (userInput.startsWith("deadline ") || userInput.equals("deadline")) {
+            String[] deadlineAttributes = validateDeadlineInput(userInput);
+            Task newTask = tasks.addDeadlineToList(deadlineAttributes[0], deadlineAttributes[1]);
+            response = ui.getNewTaskMessage(newTask, tasks.size());
+        } else if (userInput.startsWith("event ") || userInput.equals("event")) {
+            String[] eventAttributes = validateEventInput(userInput);
+            Task newTask = tasks.addEventToList(eventAttributes[0], eventAttributes[1], eventAttributes[2]);
+            response = ui.getNewTaskMessage(newTask, tasks.size());
+        } else if (userInput.startsWith("delete ") || userInput.equals("delete")) {
+            int listIndex = validateListInput(userInput, "delete", tasks.size());
+            Task deletedTask = tasks.deleteTask(listIndex);
+            response = ui.getDeletedTaskMessage(deletedTask, tasks.size());
+        } else if (userInput.startsWith("find ") || userInput.equals("find")) {
+            String keyword = validateFindInput(userInput);
+            ArrayList<Task> keywordTasks = tasks.findKeyword(keyword);
```
Perhaps this would be better written as a case statement, or with added line breaks for readability?

CaptainKeqing commented on src/main/java/toothless/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+     * @return The String indicating X if done, and nothing if not.
+     */
+    public String getStatusIcon() {
+        return (this.isDone ? "X" : " "); // mark done duke.task with X
```
Is this comment variable name a relic of intellij refactor?

CaptainKeqing commented on src/main/java/toothless/ui/MainWindow.java:
```diff
@@ -0,0 +1,73 @@
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
```
Should there be a newline here to separate logical import groups?

CaptainKeqing commented on src/main/java/toothless/storage/Storage.java:
```diff
@@ -0,0 +1,108 @@
+package toothless.storage;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Should these imports be separated by a new line for import groups?

CaptainKeqing reviewed (commented):
LGTM. Just a few nits generally related to Javadocs and imports.

getsquared committed:
Improve DialogBox GUI and Fix Text Truncation

getsquared committed:
Improve Code Quality

Review comments from first PR rectified, including naming standards for
variables and methods, and JavaDoc formatting.

A new function addTaskToTaskList() was abstracted from loadStorage() in
Storage.java to shorten loadStorage() method and fix deep nesting.

In Parser.java:
parseInput() method was made shorter by abstracting out code for each
case to individual parse functions.

Todo:
* Fix the tests for Parser.java

getsquared committed:
Update ParserTest and Fix Parser Bugs

Update tests for parseEventInput to new method signature.

parseEventInput method in Parser matches(regex) allows extra flags.

Introduces weird behavior like accepting /from /from /to, and using
dates from the two /from flags.

parseEventInput method in Parser split(regex) splits attributes
wrongly due to strip(), causing /from attribute to be in description.

Similar bugs are seen in parseDeadlineInput.

Let's
* Update matches(regex) under parseEventInput and parseDeadlineInput
to reject extra flags.
* Update strip(regex) under parseEventInput and parseDeadlineInput to
split attributes correctly.

getsquared committed:
Fix DialogBox Text Wrap Bug

getsquared committed:
Set up Gradle CI Workflow

getsquared committed:
Add assertions to TaskList.java

getsquared committed:
Merge pull request #1 from getsquared/branch-A-CodeQuality

Improve Code Quality

getsquared committed:
Revert "Merge pull request #1 from getsquared/branch-A-CodeQuality"

This reverts commit e086f5f0211af5e67192aa2d8bea836abffe24de, reversing
changes made to 29353e708eba42f39324c92e5124ffcf0786bbf5.

getsquared committed:
Merge pull request #2 from getsquared/branch-A-Assertions

Add Assertions

getsquared committed:
Merge branch 'master' into branch-A-Assertions

getsquared committed:
Merge branch 'master' into branch-A-CodeQuality

getsquared committed:
Add tagging to tasks

getsquared committed:
Merge branch 'branch-C-Tagging'

getsquared committed:
Merge branch 'branch-A-Assertions'

getsquared committed:
Merge branch 'branch-A-CodeQuality'

getsquared committed:
Fix ParserTest after weird merge conflict

getsquared committed:
Improve Chat DialogBox GUI

getsquared committed:
Merge branch 'branch-A-BetterGui'

getsquared committed:
Update error responses

getsquared committed:
Add User Guide and Update README

getsquared committed:
Merge branch 'branch-A-UserGuide'

getsquared committed:
Update User Guide

getsquared committed:
Edit Ui.png
